lisa/transformers/upgrade_packages.py

```diff
@@ -0,0 +1,215 @@
+from dataclasses import dataclass, field
+from typing import Any, Dict, List, Optional, Type
+
+from dataclasses_json import dataclass_json
+
+from lisa import schema
+from lisa.node import Node, quick_connect
+from lisa.operating_system import Debian, Ubuntu
+from lisa.tools import Uname
+from lisa.transformer import Transformer
+from lisa.util import field_metadata, subclasses
+from lisa.util.logger import Logger, get_logger
+
+
+@dataclass_json()
+@dataclass
+class UpgradeInstallerSchema(schema.TypedSchema, schema.ExtendableSchemaMixin):
+    ...
+
+
+@dataclass_json
+@dataclass
+class UpgradeInstallerTransformerSchema(schema.Transformer):
+    # SSH connection information to the node
+    connection: Optional[schema.RemoteNode] = field(
+        default=None, metadata=field_metadata(required=True)
+    )
+    repo_updater: Optional[UpgradeInstallerSchema] = field(
```

Comment:
haven't seen the reason to split repo updater and upgrade into two parts.  so the two repo_updater and installer are not necessary.

Reply:
The pipeline will have two stages, one where the transformer only changed the repo source than tests will be run (such as the dns servicing test) which updates packages and tests functionality without a reboot.

The second stage will update all packages then run a full LISA validation.

Since it is a common scenario for customers to have package updates without rebooting the VM, we want to test that functionality as well. Splitting up the repo update and the package upgrade will help us test these two different scenarios.

Reply:
I can split this into two transformers, I just thought that updating the repo list and upgrading packages was closely related. I imagine that both stages could have different implementations. You might want to remove repos, add a custom repo, add a ppa repo. You might want to upgrade specific packages, remove packages, etc.

Whether or not a reboot is done at the end of the transformer doesn't really matter, because the image is exported as a vhd and then rebooted. 
The only way to test the "upgrade and don't reboot" scenario is to have the upgrades be part of the test. As far as I know, we have no way to ensure a clean environment after every test, and to restrict the tests to ones that don't perform a reboot in the test.